### PR TITLE
move setup code from startControllers into the config

### PR DESCRIPTION
**Pull request by stewart-yu**
**What this PR does / why we need it**:
We should follow the apiserver example, and move more setup code from startControllers into the config

parts of PR [https://github.com/kubernetes/kubernetes/issues/59483](https://github.com/kubernetes/kubernetes/issues/59483)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/cloud-controller-manager/app/config/config.go [hunk 1]
```diff
@@ -46,6 +48,12 @@ type Config struct {
 
 	// the event sink
 	EventRecorder record.EventRecorder
+
+	// ClientBuilder will provide a client for this controller to use
+	ClientBuilder controller.ControllerClientBuilder
+
+	// InformerFactory gives access to informers for the controller.
+	InformerFactory informers.SharedInformerFactory
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/config/config.go]: would keep the name "SharedInformers". It says more about the use of this.

File: cmd/cloud-controller-manager/app/options/options.go [hunk 2]
```diff
@@ -222,16 +210,69 @@ func (o *CloudControllerManagerOptions) Validate() error {
...
+		}
+	} else {
+		c.ClientBuilder = rootClientBuilder
+	}
+	versionedClient := rootClientBuilder.ClientOrDie("shared-informers")
+	c.InformerFactory = informers.NewSharedInformerFactory(versionedClient, resyncPeriod(c)())
+
+	// If apiserver is not running we should wait for some time and fail only then. This is particularly
+	// important when we start apiserver and controller manager at the same time.
+	err = genericcontrollermanager.WaitForAPIServer(versionedClient, 10*time.Second)
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/options/options.go]: this does not belong here IMO. This is plumbing code and shouldn't have anything about synchronization with other components.

File: cmd/cloud-controller-manager/app/options/options.go [hunk 3]
```diff
@@ -222,16 +210,69 @@ func (o *CloudControllerManagerOptions) Validate() error {
...
+
 // Config return a cloud controller manager config objective
 func (o *CloudControllerManagerOptions) Config() (*cloudcontrollerconfig.Config, error) {
 	if err := o.Validate(); err != nil {
 		return nil, err
 	}
 
-	c := &cloudcontrollerconfig.Config{}
-	if err := o.ApplyTo(c, CloudControllerManagerUserAgent); err != nil {
+	kubeconfig, err := clientcmd.BuildConfigFromFlags(o.Master, o.Kubeconfig)
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/options/options.go]: I don't like dependencies of the config on the options in this func. Logically, this `Config` func does not belong into the options at all. All the dependencies between options and config should be part of `ApplyTo`.

File: cmd/cloud-controller-manager/app/options/options.go (hunk 2 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/options/options.go]: add `VersionedClient ` to restore `versionedClient` in config struct, then keep origin.

**Comment by sttts** [comment]
@andrewsykim thanks. Looking at it now.

File: cmd/cloud-controller-manager/app/options/options.go [hunk 4]
```diff
@@ -146,40 +150,35 @@ func (o *CloudControllerManagerOptions) AddFlags(fs *pflag.FlagSet) {
 
 // ApplyTo fills up cloud controller manager config with options.
 func (o *CloudControllerManagerOptions) ApplyTo(c *cloudcontrollerconfig.Config, userAgent string) error {
-	if err := o.CloudProvider.ApplyTo(&c.ComponentConfig.CloudProvider); err != nil {
+	var err error
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/options/options.go]: the old style is usually what we use in Kube.

**Comment by sttts** [comment]
Structure looks great now.

/lgtm

**Comment by andrewsykim** [comment]
/approve

thanks @stewart-yu!

**Pull request merged**